rfd/00XY-session-locking.md

```diff
@@ -0,0 +1,268 @@
+    string Node;
+
+    // MFADevice specifies the ID of an MFA device recorded in a user certificate.
+    string MFADevice;
+}
+```
+
+In most cases `LockTarget` should list just a single attribute.
+(It would be preferable to define it as a `oneof`, however setting a custom
+`jsontag` for `oneof` is [not supported by `gogoproto`](https://github.com/gogo/protobuf/issues/623).)
+
+Note that `Lock` is not a singleton resource: when there are multiple
+`Lock`s stored (and in force), it suffices for an interaction to be matched
+by any one of the locks to be terminated/disabled.
+
+#### Relation to `User.Status`
+
+There already exists a field of `User` resource that is used to
+capture a user lock in connection with failed Web UI login attempts.
+
+This `Status` field and its `LoginStatus` definition are superseded by
+`Lock`.  All of its use cases should be converted to `Lock`.
+
+The `Lock` approach allows to specify locks for entities that are only
+yet to exist or exist merely transiently (such as SSO user objects).  It could
+also help with alleviating the load associated with caching `User` resources.
+
+#### `tctl` support
+
+`Lock` resources can be managed using `tctl [get|create|rm]`.  In this way,
+it is possible to update or remove a lock after it has been created.
+
+There will be a special `tctl lock` helper provided, to facilitate creation
+of new `Lock`s, see Scenarios below.
+
+### Propagation within a Teleport cluster
+
+For the purposes of lock detection and evaluation, the `Lock` resource will be propagated using the cache system to all Teleport roles/services.
+
+An `auth.AccessPoint` wrapper will be introduced that makes the getters return a `trace.ConnectionProblem` for stale data. The wrapper will be parametrized by a duration that defines the maximum failure period after which the data are considered stale.
+
+`LockTarget` will be convertible into a filter for `types.WatchKind`, allowing
+the creation of watchers that match specific contexts. The `GetLocks` method
+should also accept `LockTarget`s as filters.
+
+### Fallback mode
+
+If the "strict" cache returns an error indicating stale data, there is a decision to be made about whether to rely on the last known locks. There will be two levels on which to determine this decision.
+
+1. If a lockable transaction involves a user and the user's certificate has an extension named `teleport-lock-fallback`, the transaction is locked out based on the extension's value which comes from the user's RBAC role:
+
+```
+spec:
+   options:
+       # In strict mode, if the locks are not up to date, all matching interactions will be terminated.
+       # In best_effort mode, the most recent view of the locks will remain to be used.
+       lock: [strict|best_effort]
+```
+
+2. A `LockFallback` field is added to the `ClusterAuthPreference` resource. The new field configures the cluster-wide fallback mode that applies when a more specific hint is not available. It defaults to `strict`.
```

Comment:
Defaulting to strict breaks backwards compatibility, since `strict` mode significantly reduces HA.  I think we should default to `best_effort` instead, since that is less likely to cause an unexpected change in behavior for folks that do not want/need this feature.

Reply:
Agree with @fspmarshall . We can set default strict because it will break backwards compatibility

Reply:
No problem, changed. I got the impression `strict` should be the default from https://github.com/gravitational/teleport/pull/7286#issuecomment-862624598:
> By default, in a secure system, all sessions could be strict.`